Add Agones Kubernetes API docs generator

aLekSer opened the pull request:
Add [gen-crd-api-reference-docs](https://github.com/ahmetb/gen-crd-api-reference-docs) utility into build-image, generate Agones CRD reference html document, map this doc to existent docs structure.
Covers #409.

aLekSer commented:
This is an initial commit. Some changes to the templates could occur in the future. Current look and feel:


<img width="990" alt="Agones Kubernetes CRD docs" src="https://user-images.githubusercontent.com/5626814/54212053-68f2e000-44f3-11e9-9607-dcb9035ecce9.png">

aLekSer commented:
Issue with resulting file position in Cloud Build:
```
Output to a file /go/src/agones.dev/agones/site/content/en/docs/Reference/agones_crd_api_reference.html
diff /workspace/site/content/en/docs/Reference/agones_crd_api_reference.html /tmp/generated.html
```
Locally I have for `test-gen-api-docs: expected_docs := $(site_path)/$(REL_PATH)` :
```
diff /Users/alexander.apalikov/go/src/agones.dev/agones/site/content/en/docs/Reference/agones_crd_api_reference.html /tmp/generated.html
```

aLekSer commented:
Found an issue, I have added the special kubebuilder declaration in `/pkg/apis/stable/v1alpha1/gameserver.go`, which is redundant, like that:
```
// +kubebuilder:resource:path=gameservers
```

markmandel commented on build/includes/website.mk:
```diff
@@ -24,6 +24,23 @@
 
+REL_PATH := content/en/docs/Reference/agones_crd_api_reference.html
+API_DOC_PATH := "$(mount_path)/site/$(REL_PATH)"
+GEN_API_DOCS ?= docker run -e FILE=$(API_DOC_PATH) --rm -i $(common_mounts) $(build_tag) bash -c "/go/src/agones.dev/agones/site/gen-api-docs.sh"
```
Curious as to the motivation to put all of this in variables? Seems like it wouldn't change?

aLekSer replied:
I have added those options to be able to split test html result and normal one. GEN_API_DOCS used twice, that's why it was extracted.

markmandel commented on build/includes/website.mk:
```diff
@@ -59,6 +76,7 @@ site-deploy-preview: site-static-preview
 
 site-test:
 	# generate actual html and run test against - provides a more accurate tests
+	$(MAKE) test-gen-api-docs
```
I don't have a good answer for this - but because the website works off master - if we change our CRD data structures, that will change the website as well, and be out of sync until next release.

Only idea I've got - that in the advent of that happening - rename the old page and put an `expiryDate` of the next release in the toml.... but how would we get a publishDate in the new one, and still pass tests?

I'm just thinking out loud here, but any ideas?

aLekSer replied:
Tried to use custom short code in front-matter and it seems that variables are not supported there.
So currently I am thinking of combining one full generated html with `expiryVersion` shortcode. And new one with `publishVersion` shortcode in the same file.
And during test just compare the second half with `publishVersion`. And during release move bottom half to the top and generate new one at the bottom.

markmandel replied:
That sounds fancy - but makes sense to me! I assume you can make that work? :smile:

aLekSer replied:
I have updated `gen-api-docs.sh` script. It now handles `base_version` environment variable.
What it does is get all lines before `publishVersion` line. Then extract what in the middle of this feature shortcode. Compares with what was generated if diff then do version compare.
Then set `expiryVersion` of old part with new `base_version` fresh generated doc now have new `publishVersion`.

aLekSer commented:
Reworked generation script it now handles the version in `publishVersion` shortcode. 
Old code becomes expiredVersion with new `base_version`. Test now compares only what is in the middle of `publishVersion` with the output of go tool [gen-crd-api-reference-docs](https://github.com/ahmetb/gen-crd-api-reference-docs).
Two calls to `awk` and one `sed` do the job.
We don't need to care about switching the versions on release process - it is automated for this file.

aLekSer commented:
I think I need to add md instructions on how to use new make target `gen-api-docs`. It was created to work in a similar way to  `make gen-install` when we call this on each crd update and just `git add` the resulted `.../Reference/agones_crd_api_reference.html`. So basically two subsequent calls of `	$(MAKE) test-gen-api-docs` would result in pass the test, but on CI if someone forgot to commit this html than test would fail.

aLekSer commented:
Fixed one issue, there was a commit-id in a default template. So I removed it, so now different git commit would generate the same html.
https://github.com/GoogleCloudPlatform/agones/pull/645/files#diff-b75745dde8c5e52ecb85780a42ac3669

markmandel commented:
Apologies - looks like I never submitted those comments :/ they've been sitting there forever. But otherwise, this is good to go.

aLekSer commented on site/content/en/docs/Guides/access-api.md:
```diff
@@ -10,7 +10,11 @@ description: >
-which can be accessed and manipulated through the Kubernetes API.
+which can be accessed and manipulated through the Kubernetes API. 
+
+{{% feature publishVersion="0.9.0" %}}
```
Yes, I can fix that version, but then I need to check that test still pass.

aLekSer commented:
Merged master, fixed publishVersion and Makefile parameter.

ahmetb commented on build/build-image/Dockerfile:
```diff
@@ -123,6 +123,11 @@ RUN mkdir /tmp/hugo && \
 
+# install API reference docs generator
+RUN mkdir -p /go/src/github.com/ahmetb && \
+    cd /go/src/github.com/ahmetb && git clone https://github.com/ahmetb/gen-crd-api-reference-docs && \
```
I think you should probably pin to a commit here.
Given there's a copy of the template in this repo, I don't want to accidentally break the downstream.

aLekSer replied:
Thanks @ahmetb for the review. I was thinking about that, will provide an update in a separate PR.
